### PR TITLE
script: make Instructions iterator enforce minimal pushes

### PR DESCRIPTION
**Pull request by apoelstra**

### CODE REVIEW

**Comment by apoelstra** [comment]
@TheBlueMatt what do you think about this? It will make script fragments with non-minimal pushes error out when you try to iterate over them.

Maybe we could make it optional?

The problem is that once something has become an `Instructions::PushBytes(..)` it's impossible to tell the original script fragment, so script parsers that use this iterator will fail to roundtrip if non-minimal pushes are allowed as input.

File: src/blockdata/script.rs [hunk 1]
```diff
@@ -151,6 +151,8 @@ display_from_debug!(Builder);
 /// would help you.
 #[derive(PartialEq, Eq, Debug, Clone)]
 pub enum Error {
+    /// Something did a non-minimal push
+    NonMinimalPush,
```
Comment (dpc) [inline src/blockdata/script.rs]: You might add  https://github.com/bitcoin/bips/blob/master/bip-0062.mediawiki#Push_operators or something to help people.

Comment (apoelstra) [reply src/blockdata/script.rs]: Sure, good link

**Comment by apoelstra** [comment]
Replaced with optional version; also fixed bug where errors would iterate forever. (This is plausibly reasonable behaviour but it's incredibly irritating because it makes `collect()` consume all your memory and take forever doing so.)

**Comment by apoelstra** [comment]
This is the last thing blocking 0.14. Matt found a non-minimal case that this code doesn't detect - the fix is quick but we're also trying to goad the fuzzer into detecting it. Should be done by end of day.

File: src/blockdata/script.rs [hunk 2]
```diff
@@ -373,6 +377,17 @@ impl Script {
                                opcodes::All::from(self.0[0]).classify() == opcodes::Class::IllegalOp)
     }
 
+    /// Iterate over the script in the form of `Instruction`s, which are an enum covering
+    /// opcodes, datapushes and errors. At most one error will be returned and then the
+    /// iterator will end. To instead iterate over the script as sequence of bytes, treat
+    /// it as a slice using `script[..]` or convert it to a vector using `into_vec()`.
```
Comment (jeandudey) [inline src/blockdata/script.rs]: This doc needs to be updated, `into_vec` was removed.

**Comment by apoelstra** [comment]
Fixed doc, rebased, updated fuzzer to check for nonminimal numbers.

**Comment by apoelstra** [comment]
Fixed fuzzer to not attempt to replace `00` with `OP_0` :}

**Pull request merged**